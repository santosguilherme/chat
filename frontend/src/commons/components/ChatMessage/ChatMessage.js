import React from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';

import {Message, MessageRow} from './styles/Container';
import {MessageContent, MessageTime} from './styles/Content';

function ChatMessage({align, backgroundColor, userName, message, dateTime}) {
  const locale = 'en-US'; // TOOD: receber vindo do settings
  const messageTime = new Date(dateTime).toLocaleTimeString(locale, {
    timeStyle: 'short',
    hour12: false, // TOOD: receber vindo do settings
  });

  return (
    <MessageRow left={align === 'left'}>
      <Message backgroundColor={backgroundColor}>
        <MessageContent>
          <Typography variant="subtitle2">{userName}</Typography>
          <Typography variant="body2">{message}</Typography>
        </MessageContent>
        <MessageTime>
          <Typography variant="caption">{messageTime}</Typography>
        </MessageTime>
      </Message>
    </MessageRow>
  );
}

ChatMessage.defaultProps = {
  align: 'right',
};

ChatMessage.propTypes = {
  align: PropTypes.oneOf(['left', 'right']),
  backgroundColor: PropTypes.string,
  userName: PropTypes.string,
  message: PropTypes.string,
  dateTime: PropTypes.string,
};

export default ChatMessage;
