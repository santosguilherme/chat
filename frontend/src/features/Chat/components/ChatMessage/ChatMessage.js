import React from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';

import {Message, MessageRow} from './styles/Container';
import {MessageContent, MessageTime} from './styles/Content';
import {Tail} from './styles/Tail';
import {FormattedTime, useIntl} from 'react-intl';

function ChatMessage({align, backgroundColor, userName, message, dateTime, hour12}) {
  const intl = useIntl();
  const date = new Date(dateTime);

  return (
    <MessageRow left={align === 'left'}>
      <Message backgroundColor={backgroundColor} align={align}>
        <Tail backgroundColor={backgroundColor} align={align} />
        <MessageContent>
          <Typography variant="subtitle2">{userName}</Typography>
          <Typography variant="body2">{message}</Typography>
        </MessageContent>
        <MessageTime>
          <Typography variant="caption" title={date.toLocaleString(intl.locale)}>
            <FormattedTime value={date} hour12={hour12} />
          </Typography>
        </MessageTime>
      </Message>
    </MessageRow>
  );
}

ChatMessage.defaultProps = {
  align: 'left',
  backgroundColor: 'background.paper',
};

ChatMessage.propTypes = {
  align: PropTypes.oneOf(['left', 'right']),
  backgroundColor: PropTypes.string,
  userName: PropTypes.string,
  message: PropTypes.string,
  dateTime: PropTypes.string,
  hour12: PropTypes.bool.isRequired,
};

export default ChatMessage;
