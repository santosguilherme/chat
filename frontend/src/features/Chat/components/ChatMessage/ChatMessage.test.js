import React from 'react';

import { customRender } from 'commons/utils/test/reactTestingLibrary';

import ChatMessage from './ChatMessage';

function renderChatMessage(options = {}) {
  const {
    dateTime = '2020-02-18T02:36:37.071Z',
    hour12 = true,
    message = 'any message',
    userName,
    align,
    backgroundColor,
  } = options;
  return customRender(
    <ChatMessage
      dateTime={dateTime}
      hour12={hour12}
      message={message}
      userName={userName}
      align={align}
      backgroundColor={backgroundColor}
    />,
  );
}

describe('ChatMessage', () => {
  describe('hour12', () => {
    test('shows the time in 12 hours format', () => {
      const { getByText } = renderChatMessage({ dateTime: '2020-02-18T02:30:37.071Z' });

      expect(getByText('11:30 PM')).toBeInTheDocument();
    });

    test('shows the time in 24 hours format', () => {
      const { getByText } = renderChatMessage({ hour12: false, dateTime: '2020-02-18T02:30:37.071Z' });

      expect(getByText('23:30')).toBeInTheDocument();
    });
  });

  describe('message', () => {
    test('is displayed', () => {
      const { getByText } = renderChatMessage({ message: 'my chat message' });

      expect(getByText('my chat message')).toBeInTheDocument();
    });
  });

  describe('userName', () => {
    test('is displayed when informed', () => {
      const { getByText } = renderChatMessage({ userName: 'any-user' });

      expect(getByText('any-user')).toBeInTheDocument();
    });

    test('is not displayed when omitted', () => {
      const { container } = renderChatMessage();

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('align', () => {
    test('the default value', () => {
      const { container } = renderChatMessage();

      expect(container.firstChild).toMatchSnapshot();
    });

    test('the custom value', () => {
      const { container } = renderChatMessage({ align: 'right' });

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('backgroundColor', () => {
    test('the default value', () => {
      const { container } = renderChatMessage();

      expect(container.firstChild).toMatchSnapshot();
    });

    test('the custom value', () => {
      const { container } = renderChatMessage({ backgroundColor: 'chat.loggedMessage' });

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
