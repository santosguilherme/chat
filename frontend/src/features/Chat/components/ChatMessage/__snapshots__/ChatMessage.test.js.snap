// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessage align the custom value 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 1rem;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c1 {
  display: grid;
  grid-template-columns: auto auto;
  box-shadow: 0 1px 0.5px rgba(0,0,0,.13);
  position: relative;
  border-radius: 7.5px;
  background-color: #fff;
  border-top-right-radius: 0;
}

.c3 {
  padding: 0.7rem;
}

.c4 {
  -webkit-align-self: end;
  -ms-flex-item-align: end;
  align-self: end;
  padding: 0.5rem;
}

.c2 {
  position: absolute;
  width: 0;
  height: 0;
  border-top: 10px solid #fff;
  border-right: 10px solid transparent;
  right: -10px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <span
      class="c2"
    />
    <div
      class="c3"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      />
      <p
        class="MuiTypography-root MuiTypography-body2"
      >
        any message
      </p>
    </div>
    <div
      class="c4"
    >
      <span
        class="MuiTypography-root MuiTypography-caption"
        title="2/17/2020, 11:36:37 PM"
      >
        11:36 PM
      </span>
    </div>
  </div>
</div>
`;

exports[`ChatMessage align the default value 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 1rem;
}

.c1 {
  display: grid;
  grid-template-columns: auto auto;
  box-shadow: 0 1px 0.5px rgba(0,0,0,.13);
  position: relative;
  border-radius: 7.5px;
  background-color: #fff;
  border-top-left-radius: 0;
}

.c3 {
  padding: 0.7rem;
}

.c4 {
  -webkit-align-self: end;
  -ms-flex-item-align: end;
  align-self: end;
  padding: 0.5rem;
}

.c2 {
  position: absolute;
  width: 0;
  height: 0;
  border-top: 10px solid #fff;
  border-left: 10px solid transparent;
  left: -10px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <span
      class="c2"
    />
    <div
      class="c3"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      />
      <p
        class="MuiTypography-root MuiTypography-body2"
      >
        any message
      </p>
    </div>
    <div
      class="c4"
    >
      <span
        class="MuiTypography-root MuiTypography-caption"
        title="2/17/2020, 11:36:37 PM"
      >
        11:36 PM
      </span>
    </div>
  </div>
</div>
`;

exports[`ChatMessage backgroundColor the custom value 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 1rem;
}

.c1 {
  display: grid;
  grid-template-columns: auto auto;
  box-shadow: 0 1px 0.5px rgba(0,0,0,.13);
  position: relative;
  border-radius: 7.5px;
  background-color: #DCF8C6;
  border-top-left-radius: 0;
}

.c3 {
  padding: 0.7rem;
}

.c4 {
  -webkit-align-self: end;
  -ms-flex-item-align: end;
  align-self: end;
  padding: 0.5rem;
}

.c2 {
  position: absolute;
  width: 0;
  height: 0;
  border-top: 10px solid #DCF8C6;
  border-left: 10px solid transparent;
  left: -10px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <span
      class="c2"
    />
    <div
      class="c3"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      />
      <p
        class="MuiTypography-root MuiTypography-body2"
      >
        any message
      </p>
    </div>
    <div
      class="c4"
    >
      <span
        class="MuiTypography-root MuiTypography-caption"
        title="2/17/2020, 11:36:37 PM"
      >
        11:36 PM
      </span>
    </div>
  </div>
</div>
`;

exports[`ChatMessage backgroundColor the default value 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 1rem;
}

.c1 {
  display: grid;
  grid-template-columns: auto auto;
  box-shadow: 0 1px 0.5px rgba(0,0,0,.13);
  position: relative;
  border-radius: 7.5px;
  background-color: #fff;
  border-top-left-radius: 0;
}

.c3 {
  padding: 0.7rem;
}

.c4 {
  -webkit-align-self: end;
  -ms-flex-item-align: end;
  align-self: end;
  padding: 0.5rem;
}

.c2 {
  position: absolute;
  width: 0;
  height: 0;
  border-top: 10px solid #fff;
  border-left: 10px solid transparent;
  left: -10px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <span
      class="c2"
    />
    <div
      class="c3"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      />
      <p
        class="MuiTypography-root MuiTypography-body2"
      >
        any message
      </p>
    </div>
    <div
      class="c4"
    >
      <span
        class="MuiTypography-root MuiTypography-caption"
        title="2/17/2020, 11:36:37 PM"
      >
        11:36 PM
      </span>
    </div>
  </div>
</div>
`;

exports[`ChatMessage userName is not displayed when omitted 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 1rem;
}

.c1 {
  display: grid;
  grid-template-columns: auto auto;
  box-shadow: 0 1px 0.5px rgba(0,0,0,.13);
  position: relative;
  border-radius: 7.5px;
  background-color: #fff;
  border-top-left-radius: 0;
}

.c3 {
  padding: 0.7rem;
}

.c4 {
  -webkit-align-self: end;
  -ms-flex-item-align: end;
  align-self: end;
  padding: 0.5rem;
}

.c2 {
  position: absolute;
  width: 0;
  height: 0;
  border-top: 10px solid #fff;
  border-left: 10px solid transparent;
  left: -10px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <span
      class="c2"
    />
    <div
      class="c3"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      />
      <p
        class="MuiTypography-root MuiTypography-body2"
      >
        any message
      </p>
    </div>
    <div
      class="c4"
    >
      <span
        class="MuiTypography-root MuiTypography-caption"
        title="2/17/2020, 11:36:37 PM"
      >
        11:36 PM
      </span>
    </div>
  </div>
</div>
`;
